daemon/response.go

```diff
@@ -281,55 +283,163 @@ func (f fileResponse) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		// condition for the read loop, and then do the error handling in
+		// the main go routine.
+		if err := decoder.Decode(&log); err != nil {
+			// Ignore the return value here as we are breaking out
+			// anyway and not sending more messages.
+			rr.safeSendError(e, err)
 			break
 		}
 
-		writer.WriteByte(0x1E) // RS -- see ascii(7), and RFC7464
-
-		// ignore the error...
-		t, _ := log.Time()
-		if err = enc.Encode(client.Log{
-			Timestamp: t,
-			Message:   log.Message(),
-			SID:       log.SID(),
-			PID:       log.PID(),
-		}); err != nil {
+		if closed := rr.safeSendLog(c, log); closed {
 			break
 		}
+	}
+}
+
+func (rr *journalLineReaderSeqResponse) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json-seq")
+
+	flusher, hasFlusher := w.(http.Flusher)
+	writer := bufio.NewWriter(w)
+	enc := json.NewEncoder(writer)
+
+	// Buffer 128 (arbitrary, seems appropriate) messages, and
+	// the number of readers in errors as we know exactly how
+	// many errors to expect on the channel.
+	c := make(chan systemd.Log, 128)
+	e := make(chan error, len(rr.readers))
+	for _, r := range rr.readers {
+		go rr.logReader(r, c, e)
+	}
+
+	writeError := func(err error) {
+		// RS -- see ascii(7), and RFC7464
+		fmt.Fprintf(writer, `\x1E{"error": %q}\n`, err)
+		logger.Noticef("cannot stream response; problem reading: %v", err)
+	}
+
+	writeLogs := func(logs []systemd.Log) error {
+		// sort by timestamp ascending
+		sort.Slice(logs, func(i, j int) bool {
+			ti, _ := logs[i].Time()
+			tj, _ := logs[j].Time()
+			return ti.Before(tj)
+		})
+
+		for _, l := range logs {
+			writer.WriteByte(0x1E) // RS -- see ascii(7), and RFC7464
+
+			// ignore the error...
+			t, _ := l.Time()
+			if err := enc.Encode(client.Log{
+				Timestamp: t,
+				Message:   l.Message(),
+				SID:       l.SID(),
+				PID:       l.PID(),
+			}); err != nil {
+				return err
+			}
+		}
 
 		if rr.follow {
-			if e := writer.Flush(); e != nil {
-				break
+			if err := writer.Flush(); err != nil {
+				return errCannotWriteToClient
 			}
 			if hasFlusher {
 				flusher.Flush()
 			}
 		}
+		return nil
 	}
-	if err != nil && err != io.EOF {
-		fmt.Fprintf(writer, `\x1E{"error": %q}\n`, err)
-		logger.Noticef("cannot stream response; problem reading: %v", err)
+
+	var logReadersDone int
+	for {
+		var logs []systemd.Log
+
+		// Now we spend a small amount of time batch reading all available logs
+		// on the log/error channel, the reason we do this is to make sure we read
+		// everything available for the initial/final batch of logs (when following) or
+		// we read everything available when not following. It's done like this to
+		// ensure that output is consistent and sorted correctly by timestamp when
+		// multiple log-sources are in play. The timeout should be small enough so
+		// the user won't notice the speed of logs appearing on the stream.
+		timeout := time.After(time.Millisecond * 25)
+		for {
```

Comment:
it's a bit more idiomatic to do:

terminate := false
for !terminate {
  ... terminate = true
}